File: src/converters/ucmuon_to_phits.f90
!=============================================================================
! ucmuon_to_phits.f90  —  UCMuon → PHITS s-type=17 dump converter
! UCLouvain Muography Group | Hamid Basiri <hamid.basiri@uclouvain.be>
!
! Fast compiled converter — handles GB-scale files efficiently.
! Reads from stdin, writes to stdout (pipe-friendly).
!
! Usage
! -----
!   ./ucmuon_to_phits gen       < ucmuon_selected.dat   > output_phits.dat
!   ./ucmuon_to_phits transport < ucmuon_underground.dat > output_phits.dat
!
! Or via the SLURM scripts (called automatically after each job).
!
! Input formats
! -------------
!   gen mode       : 13-col or 14-col generator output (auto-detected)
!   transport mode : 18-col transport output (alive==1 rows only)
!
! Output format  (PHITS s-type=17 ASCII dump)
! --------------------------------------------
!   10 columns per muon, D-exponent notation (Fortran 1pd24.15):
!     kf  x[cm]  y[cm]  z[cm]  u  v  w  Ekin[MeV]  wt  time[ns]
!
!   kf codes (PDG): mu- = 13,  mu+ = -13
!
!   Use in PHITS input:
!     [Source]
!     s-type = 17
!     file   = ucmuon_selected_phits.dat
!     dump   = -10
!     1 2 3 4 5 6 7 8 9 10
!
! Performance
! -----------
!   Fortran formatted I/O on a 1GB file: ~15-30 s (vs ~100 s in Python).
!   No MPI or OMP needed — purely serial post-processing.
!=============================================================================
program ucmuon_to_phits
  implicit none

  real(8), parameter :: MMUON_GEV = 0.105658370d0
  real(8), parameter :: GEV_TO_MEV = 1000.0d0

  character(16)  :: mode_arg
  character(512) :: linebuf
  integer :: ios, ncols, nwritten, nskipped

  ! Generator columns
  integer  :: id_g, charge_g, hit_flag_g, det_mask_g
  real(8)  :: x_g, y_g, z_g, p_g, px_g, py_g, pz_g
  real(8)  :: theta_g, phi_g, e_g

  ! Transport columns
  integer  :: id_t, charge_t, alive_t
  real(8)  :: x_srf, y_srf, z_srf, e_srf, theta_srf, phi_srf
  real(8)  :: x_ug, y_ug, z_ug, e_ug, cx_ug, cy_ug, cz_ug
  real(8)  :: theta_ug, phi_ug

  ! PHITS output fields
  integer  :: kf
  real(8)  :: x, y, z, u, v, w, ekin_mev, norm

  !--- Read mode argument -------------------------------------------------------
  if (command_argument_count() < 1) then
    write(0,*) 'Usage: ucmuon_to_phits gen|transport < input.dat > output.dat'
    stop 1
  end if
  call get_command_argument(1, mode_arg)
  mode_arg = adjustl(mode_arg)

  if (trim(mode_arg) /= 'gen' .and. trim(mode_arg) /= 'transport') then
    write(0,*) 'ERROR: mode must be "gen" or "transport", got: ', trim(mode_arg)
    stop 1
  end if

  nwritten = 0
  nskipped = 0
  ncols    = 0

  !===========================================================================
  ! GENERATOR MODE  (13 or 14 columns)
  !===========================================================================
  if (trim(mode_arg) == 'gen') then

    do
      read(*, '(A512)', iostat=ios) linebuf
      if (ios /= 0) exit
      linebuf = adjustl(linebuf)
      if (len_trim(linebuf) == 0) cycle
      if (linebuf(1:1) == '#')   cycle

      ! Auto-detect column count from first data line
      if (ncols == 0) then
        read(linebuf, *, iostat=ios) id_g, x_g, y_g, z_g, p_g, &
              px_g, py_g, pz_g, theta_g, phi_g, e_g, charge_g, hit_flag_g, det_mask_g
        if (ios == 0) then
          ncols = 14
        else
          read(linebuf, *, iostat=ios) id_g, x_g, y_g, z_g, p_g, &
                px_g, py_g, pz_g, theta_g, phi_g, e_g, charge_g, det_mask_g
          if (ios == 0) then
            ncols = 13
          else
            write(0,*) 'ERROR: cannot parse first data line as 13 or 14 columns.'
            stop 1
          end if
        end if
      else
        if (ncols == 14) then
          read(linebuf, *, iostat=ios) id_g, x_g, y_g, z_g, p_g, &
                px_g, py_g, pz_g, theta_g, phi_g, e_g, charge_g, hit_flag_g, det_mask_g
        else
          read(linebuf, *, iostat=ios) id_g, x_g, y_g, z_g, p_g, &
                px_g, py_g, pz_g, theta_g, phi_g, e_g, charge_g, det_mask_g
        end if
        if (ios /= 0) then
          nskipped = nskipped + 1
          cycle
        end if
      end if

      ! Direction cosines from momentum components
      if (p_g > 0.0d0) then
        u = px_g / p_g
        v = py_g / p_g
        w = pz_g / p_g
      else
        u = 0.0d0;  v = 0.0d0;  w = -1.0d0
      end if

      ! Normalise (safety)
      norm = sqrt(u*u + v*v + w*w)
      if (norm > 1.0d-14) then
        u = u/norm;  v = v/norm;  w = w/norm
      end if

      ! PDG kf: mu+ = -13, mu- = +13
      kf = merge(-13, 13, charge_g == 1)

      ! Kinetic energy in MeV
      ekin_mev = max(0.0d0, (e_g - MMUON_GEV) * GEV_TO_MEV)

      ! Write PHITS record
      write(*, '(10(1pd24.15))') dble(kf), x_g, y_g, z_g, &
                                  u, v, w, ekin_mev, 1.0d0, 0.0d0
      nwritten = nwritten + 1
    end do

  !===========================================================================
  ! TRANSPORT MODE  (18 columns, alive==1 only)
  !===========================================================================
  else

    do
      read(*, '(A512)', iostat=ios) linebuf
      if (ios /= 0) exit
      linebuf = adjustl(linebuf)
      if (len_trim(linebuf) == 0) cycle
      if (linebuf(1:1) == '#')   cycle

      read(linebuf, *, iostat=ios) &
        id_t, x_srf, y_srf, z_srf, e_srf, theta_srf, phi_srf, charge_t, &
        alive_t, x_ug, y_ug, z_ug, e_ug, cx_ug, cy_ug, cz_ug, theta_ug, phi_ug

      if (ios /= 0) then
        nskipped = nskipped + 1
        cycle
      end if

      ! Skip stopped muons
      if (alive_t /= 1) then
        nskipped = nskipped + 1
        cycle
      end if

      u = cx_ug;  v = cy_ug;  w = cz_ug

      ! Normalise direction
      norm = sqrt(u*u + v*v + w*w)
      if (norm > 1.0d-14) then
        u = u/norm;  v = v/norm;  w = w/norm
      end if

      kf = merge(-13, 13, charge_t == 1)
      ekin_mev = max(0.0d0, (e_ug - MMUON_GEV) * GEV_TO_MEV)

      write(*, '(10(1pd24.15))') dble(kf), x_ug, y_ug, z_ug, &
                                  u, v, w, ekin_mev, 1.0d0, 0.0d0
      nwritten = nwritten + 1
    end do

  end if

  !===========================================================================
  ! SUMMARY  (to stderr so it doesn't pollute the PHITS file)
  !===========================================================================
  write(0, '(A,I12)') '  ucmuon_to_phits: written  ', nwritten
  if (nskipped > 0) &
    write(0, '(A,I12)') '  ucmuon_to_phits: skipped  ', nskipped
  write(0, '(A)')      '  Use in PHITS: s-type=17, dump=-10, 1 2 3 4 5 6 7 8 9 10'

  stop
end program ucmuon_to_phits

File: src/converters/ucmuon_to_phits.py
#!/usr/bin/env python3
"""
ucmuon_to_phits.py  —  UCMuon → PHITS s-type=17 dump converter
UCLouvain Muography Group | Hamid Basiri <hamid.basiri@uclouvain.be>

Converts UCMuon output files to PHITS ASCII dump format (s-type=17).
Works on both generator output (ucmuon_selected.dat) and transport
output (ucmuon_underground.dat).

Usage
-----
  python3 ucmuon_to_phits.py <input.dat> [output.dat] [--mode gen|transport]

  mode=gen        (default) reads ucmuon_selected.dat / ucmuon_surface.dat
                  13-col or 14-col generator format
                  → uses surface position and direction

  mode=transport  reads ucmuon_underground.dat (18-col transport output)
                  → uses underground position and direction, alive==1 only

Output format
-------------
  PHITS s-type=17 ASCII dump, 10 columns per muon:
    kf   x[cm]  y[cm]  z[cm]  u  v  w  Ekin[MeV]  wt  time[ns]

  kf codes (PDG): mu- = 13,  mu+ = -13

  Use in PHITS input:
    [Source]
    s-type = 17
    file   = ucmuon_phits.dat
    dump   = -10
    1 2 3 4 5 6 7 8 9 10

Column layout reference
-----------------------
  Generator 13-col: id  x  y  z  p  px  py  pz  theta  phi  E  charge  det_mask
  Generator 14-col: id  x  y  z  p  px  py  pz  theta  phi  E  charge  hit_flag  det_mask
  Transport 18-col: id  x_srf  y_srf  z_srf  E_srf  theta_srf  phi_srf  charge
                    alive  x_ug  y_ug  z_ug  E_ug  cx_ug  cy_ug  cz_ug
                    theta_ug  phi_ug
"""

import sys
import os
import math
import argparse

MUON_MASS_GEV = 0.105658370   # GeV/c²
MUON_MASS_MEV = 105.658370    # MeV/c²


def detect_format(line):
    """Return (ncols, mode) from the first data line."""
    parts = line.split()
    n = len(parts)
    if n == 13:
        return 13, 'gen'
    elif n == 14:
        return 14, 'gen'
    elif n == 18:
        return 18, 'transport'
    else:
        raise ValueError(f"Unrecognised column count: {n} (expected 13, 14, or 18)")


def direction_from_momentum(px, py, pz, p):
    """Return unit direction cosines (u, v, w) from momentum components."""
    if p > 0.0:
        return px / p, py / p, pz / p
    return 0.0, 0.0, -1.0   # fallback: straight down


def direction_from_angles(theta, phi):
    """Return unit direction cosines from zenith/azimuth angles [rad]."""
    st = math.sin(theta)
    return st * math.cos(phi), st * math.sin(phi), -math.cos(theta)


def energy_to_kf(charge, e_gev):
    """Return (kf, ekin_MeV). charge: +1=mu+, -1=mu-."""
    kf = -13 if charge == 1 else 13          # PDG: mu+=−13, mu−=+13
    ekin_mev = max(0.0, (e_gev - MUON_MASS_GEV) * 1000.0)
    return kf, ekin_mev


def fmt(v):
    """PHITS dump-a format: 1pd24.15 (D-style exponent, 24 wide, 15 decimals)."""
    s = f"{v:24.15E}"
    # Python uses 'E', PHITS expects 'D' — both are accepted by most PHITS versions
    return s


def convert(infile, outfile, mode):
    n_written = 0
    n_skipped = 0

    with open(infile, 'r') as fin, open(outfile, 'w') as fout:
        for raw in fin:
            line = raw.strip()
            if not line or line.startswith('#'):
                continue

            parts = line.split()
            ncols = len(parts)

            if mode == 'auto':
                mode = detect_format(line)[1]

            # ----------------------------------------------------------------
            # Generator format (13 or 14 columns)
            # ----------------------------------------------------------------
            if mode == 'gen':
                if ncols not in (13, 14):
                    continue
                try:
                    x     = float(parts[1])
                    y     = float(parts[2])
                    z     = float(parts[3])
                    p     = float(parts[4])
                    px    = float(parts[5])
                    py    = float(parts[6])
                    pz    = float(parts[7])
                    theta = float(parts[8])
                    phi   = float(parts[9])
                    e_gev = float(parts[10])
                    charge= int(parts[11])
                except (ValueError, IndexError):
                    n_skipped += 1
                    continue

                u, v, w = direction_from_momentum(px, py, pz, p)
                kf, ekin = energy_to_kf(charge, e_gev)

            # ----------------------------------------------------------------
            # Transport format (18 columns) — only alive muons
            # ----------------------------------------------------------------
            elif mode == 'transport':
                if ncols != 18:
                    continue
                try:
                    charge = int(parts[7])
                    alive  = int(parts[8])
                    x      = float(parts[9])
                    y      = float(parts[10])
                    z      = float(parts[11])
                    e_gev  = float(parts[12])
                    cx     = float(parts[13])
                    cy     = float(parts[14])
                    cz     = float(parts[15])
                except (ValueError, IndexError):
                    n_skipped += 1
                    continue

                if alive != 1:
                    n_skipped += 1
                    continue

                u, v, w = cx, cy, cz
                # Normalise direction (should already be unit but just in case)
                norm = math.sqrt(u*u + v*v + w*w)
                if norm > 1e-14:
                    u, v, w = u/norm, v/norm, w/norm

                kf, ekin = energy_to_kf(charge, e_gev)

            else:
                raise ValueError(f"Unknown mode: {mode}")

            # Write PHITS record: kf x y z u v w ekin wt time
            fout.write(
                fmt(float(kf)) +
                fmt(x) + fmt(y) + fmt(z) +
                fmt(u) + fmt(v) + fmt(w) +
                fmt(ekin) +
                fmt(1.0) +   # weight = 1
                fmt(0.0) +   # time = 0 ns
                '\n'
            )
            n_written += 1

    return n_written, n_skipped


def main():
    parser = argparse.ArgumentParser(
        description='Convert UCMuon output to PHITS s-type=17 dump')
    parser.add_argument('input',  help='Input file (ucmuon_selected.dat or ucmuon_underground.dat)')
    parser.add_argument('output', nargs='?', help='Output PHITS file (default: <input_stem>_phits.dat)')
    parser.add_argument('--mode', choices=['gen', 'transport', 'auto'],
                        default='auto',
                        help='Input format: gen=generator, transport=18-col, auto=detect')
    args = parser.parse_args()

    if not os.path.isfile(args.input):
        print(f"ERROR: input file not found: {args.input}")
        sys.exit(1)

    if args.output is None:
        stem = os.path.splitext(args.input)[0]
        args.output = stem + '_phits.dat'

    print(f"UCMuon → PHITS converter")
    print(f"  Input  : {args.input}  (mode={args.mode})")
    print(f"  Output : {args.output}")

    n_written, n_skipped = convert(args.input, args.output, args.mode)

    print(f"  Written: {n_written:,} muons")
    if n_skipped:
        print(f"  Skipped: {n_skipped:,} (stopped muons or bad lines)")
    print()
    print("Use in PHITS input section:")
    print("  [Source]")
    print("  s-type = 17")
    print(f"  file   = {args.output}")
    print("  dump   = -10")
    print("  1 2 3 4 5 6 7 8 9 10")


if __name__ == '__main__':
    main()
